content/300-guides/100-performance-and-optimization/150-connection-management/index.mdx

```diff
@@ -0,0 +1,286 @@
+
+#### Without an external connection pooler
+
+We recommend that you set the `connection_limit` to **1** if you are **not** using an external connection pooler. Each incoming request starts a short-lived Node.js process, and many concurrent functions with a high `connection_limit` can quickly **exhaust the _database_ connection limit** during a traffic spike.
+
+The following example demonstrates how to set the `connection_limit` to 1 in your connection string:
+
+<TabbedContent tabs={[<FileWithIcon text="PostgreSQL" icon="database"/>, <FileWithIcon text="MySQL" icon="database"/>]}>
+<tab>
+
+  ```
+  postgresql://USER:PASSWORD@HOST:PORT/DATABASE?connection_limit=1
+  ```
+
+</tab>
+<tab>
+
+  ```
+  mysql://USER:PASSWORD@HOST:PORT/DATABASE?connection_limit=1
+  ```
+
+</tab>
+</TabbedContent>
+
+#### With an external connection pooler
+
+If you are using an external connection pooler, use the default connection limit (`num_physical_cpus * 2 + 1`) as a starting point and then [tune the pool size](#optimizing-the-connection-pool). The external connection pooler prevents a traffic spike from overwhelming the database.
+
+### <inlinecode>PrismaClient</inlinecode> in serverless environments
+
+#### Instantiate `PrismaClient` outside the handler
+
+Instantiate `PrismaClient` [outside the scope of the function handler](https://github.com/prisma/e2e-tests/blob/5d1041d3f19245d3d237d959eca94d1d796e3a52/platforms/serverless-lambda/index.ts#L3) to increase the chances of reuse. As long as the handler remains 'warm' (in use), the connection is potentially reusable:
+
+```ts highlight=3;normal
+import { PrismaClient, Prisma } from '@prisma/client'
+
+const client = new PrismaClient()
+
+export async function handler() {
+    /* ... */
+}
+```
+
+#### Do not explicitly <inlinecode>$disconnect()</inlinecode>
+
+You [do not need to explicitly `$disconnect()`](../../../concepts/components/prisma-client/working-with-prismaclient/connection-management#calling-disconnect-explicitly) <span class="concept"></span> at the end of a function, as there is a possibility that the container might be reused. Opening a new connection takes time and slows down your function's ability to process requests.
+
+### Other serverless considerations
+
+#### Container reuse
+
+There is no guarnatee that subsequent nearby invocations of a function will hit the same container - for example, AWS can choose to create a new container at any time.
+
+Code should assume the container to be stateless and create a connection only if it does not exist - Prisma Client JS already implements this logic.
+
+#### Zombie connections
+
+Containers that are marked to be removed and are not being reused still keep a connection open and can stay in that state for some time (unknown and not documented from AWS). This can lead to sub-optimal utilization of the DB connections.
+
+Potential solutions<sup>1, 2</sup> include:
+
+* Using a lower idle connection timeout.
```

Comment:
This will not work in a serverless environment - the function is paused and you never execute any code that could close the idle connection. Only next time your serverless function wakes up to be used, you would realize "oh, this connection is super mega idle >1hr and should have been closed after 5 min - let's do that now". Now you have no more connection and need to open a new one :/

Reply:
@mhwelander wrote in Slack:

> Question about the zombie connections in serverless environments - which part is it that will not work:
>
> > This will not work in a serverless environment - the function is paused and you never execute any code that could close the idle connection.
>
> Is it the "cleaning up idle connections in a separate service" or "setting lower timeout"?
> And what does serverless-mysql implement?
>
> (I'm guessing from context cleaning up idle connections is not possible - I can see settings for zombieMaxTimeout)

Reply:
This really referred to "Using a lower idle connection timeout." (GH always displays the comment under the line it refers to, or highlights multiple lines if the comment is about multiple ones- in the main PR view this is most obvious, unfortunately not in the code view :/ )

An idle connection can only be closed if some code is running to close it. You can not set a magic "this connection will close itself after x second" when creating a connection. That is usually done by the program that is opening the connection, by running a timer that then closes it (e.g. in the connection pool - Prisma experimentally supports something like this as well since a recent release btw (undocumented as no idea if it actually works for now))

On serverless the code is paused and the environment frozen (with the open connection). So any "idle" setting can only be evaluated when the code is unfrozen again, but then it is already to late at the connection was blocked for long.

---

Now to the question you attributed to the line below:

> Is it the "cleaning up idle connections in a separate service" or "setting lower timeout"?
> And what does serverless-mysql implement?
> (I'm guessing from context cleaning up idle connections is not possible - I can see settings for zombieMaxTimeout)

serverless-mysql cleans up idle connetions after doing the actual work, so basically "cleaning up idle connections in a separate service" but without the "separate service" - at the cost of making all requests slower by the time the cleanup code needs to execute.

Reply:
Updated:https://607d3769e786c80007ec9d89--prisma2-docs.netlify.app/docs/guides/performance-and-optimization/connection-management#zombie-connections